File: DynamicPrograming/java/MaxPathSum.java
import java.util.HashMap;
import java.util.Map;

public final class MaxPathSum {
  private MaxPathSum() {
  }

  public static int maxSum(int[][] grid) {
    int rows = grid.length;
    if (rows == 0) {
      return 0;
    }
    int cols = grid[0].length;
    Map<String, Integer> memo = new HashMap<>();
    return dfs(0, 0, grid, rows, cols, memo);
  }

  private static int dfs(int r, int c, int[][] grid, int rows, int cols, Map<String, Integer> memo) {
    if (r == rows || c == cols) {
      return Integer.MIN_VALUE / 2;
    }
    if (r == rows - 1 && c == cols - 1) {
      return grid[r][c];
    }
    String key = r + "," + c;
    if (memo.containsKey(key)) {
      return memo.get(key);
    }
    int down = dfs(r + 1, c, grid, rows, cols, memo);
    int right = dfs(r, c + 1, grid, rows, cols, memo);
    int best = grid[r][c] + Math.max(down, right);
    memo.put(key, best);
    return best;
  }
}

File: DynamicPrograming/ts/maxPathSum.ts
export function maxPathSum(grid: number[][]): number {
  const memo = new Map<string, number>();

  const dfs = (r: number, c: number): number => {
    if (r === grid.length || c === grid[0].length) return Number.NEGATIVE_INFINITY;
    if (r === grid.length - 1 && c === grid[0].length - 1) return grid[r][c];

    const key = `${r},${c}`;
    if (memo.has(key)) return memo.get(key)!;

    const down = dfs(r + 1, c);
    const right = dfs(r, c + 1);
    const value = grid[r][c] + Math.max(down, right);
    memo.set(key, value);
    return value;
  };

  return dfs(0, 0);
}
